File: Leet_Code/Symmetric_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSymmetric_helper(TreeNode* p1, TreeNode* p2) {
        if (!p1 && !p2) {
            return true;
        } else if (!p1 || !p2) {
            return false;
        }
        if (p1->val != p2->val) {
            return false;
        }
        return isSymmetric_helper(p1->left, p2->right) && isSymmetric_helper(p1->right, p2->left);
    }
    
    bool isSymmetric(TreeNode* root) {
        if (!root) {
            return true;
        } return isSymmetric_helper(root->left, root->right);
    }
};

File: Leet_Code/Symmetric_Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean helper(TreeNode left, TreeNode right) {
        if (left == null && right == null) {
            return true;
        } else if (left == null || right == null) {
            return false;
        } else if (left.val == right.val) {
            return helper(left.left, right.right) & helper(left.right, right.left);
        } else {
            return false;
        }
    }

    public boolean isSymmetric(TreeNode root) {
        if (root == null) {
            return true;
        }
        TreeNode left_node = root.left;
        TreeNode right_node = root.right;
        return helper(left_node, right_node);
    }
}